File: model-client/src/main/java/org/modelix/model/persistent/CPHamtInternal.java
package org.modelix.model.persistent;

import java.util.Arrays;

public class CPHamtInternal extends CPHamtNode {

  public int bitmap;

  /**
   * SHA to CPHamtNode
   */
  public final String[] children;

  public CPHamtInternal(int bitmap, String[] childHashes) {
    this.bitmap = bitmap;
    this.children = childHashes;
  }

  @Override
  public String serialize() {
    return String.format("I/%s/%s",
            SerializationUtil.intToHex(bitmap),
            Arrays.stream(children).reduce((a, b) -> a + "," + b).orElse("")
    );
  }
}

File: model-client/src/main/kotlin/org/modelix/model/persistent/CPHamtInternal.kt
package org.modelix.model.persistent

import org.modelix.model.persistent.SerializationUtil.intToHex
import java.util.*

class CPHamtInternal(var bitmap: Int,
                     /**
                      * SHA to CPHamtNode
                      */
                     val children: Array<String?>) : CPHamtNode() {

    override fun serialize(): String {
        return String.format("I/%s/%s",
                intToHex(bitmap),
                Arrays.stream(children).reduce { a: String?, b: String? -> "$a,$b" }.orElse("")
        )
    }

}